packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -46,14 +46,24 @@ class GradleUtils {
     final Directory androidDir = project.android.hostAppGradleRoot;
     // Update the project if needed.
     // TODO(egarciad): https://github.com/flutter/flutter/issues/40460
-    migrateToR8(androidDir);
-    injectGradleWrapperIfNeeded(androidDir);
+    gradleUtils.migrateToR8(androidDir);
+    gradleUtils.injectGradleWrapperIfNeeded(androidDir);
 
     final File gradle = androidDir.childFile(
       platform.isWindows ? 'gradlew.bat' : 'gradlew',
     );
     if (gradle.existsSync()) {
       printTrace('Using gradle from ${gradle.absolute.path}.');
+
+      // If the Gradle executable doesn't have execute permission,
+      // then attempt to set it.
+      final FileStat gradleStat = gradle.statSync();
+      printTrace('Gradle mode: ${gradleStat.modeString()}.');
+      if (gradleStat.type != FileSystemEntityType.notFound &&
+          gradleStat.mode & 0x1 == 0) {
+        printTrace('Trying to give execute permission to Gradle.');
+        os.makeExecutable(gradle);
```

Comment:
I notice that the code in cache.dart for the executable engine artifacts also does a chmod on the containing directory:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/cache.dart#L730

@tvolkert what was the reason for that? Should `os.makeExecutable()` do the same, that is mark both the parent and the file a+r,a+x?

Reply:
customer:amplify was running into a case where they were installing the Flutter SDK to a shared location not owned by individual users, and they were seeing some cache directories not executable.

> Should `os.makeExecutable()` do the same

Check the call sites to see if that makes sense - I could see an argument in favor of this.  As long as it's documented to do so, sgtm.

Reply:
sounds good. `os.makeExecutable()` only sets `a+x` to the file passed to it. I will follow up with this.